# Assignment 5, Question 4: Data Exploration

**Points: 15**

In this notebook, you'll explore the clinical trial dataset using pandas selection and filtering techniques.

You'll use utility functions from `q3_data_utils` where helpful, but also demonstrate direct pandas operations.

## Setup

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import utilities from Q3
from q3_data_utils import load_data, detect_missing, filter_data

# Load the data
df = load_data('data/clinical_trial_raw.csv')
print(f"Loaded {len(df)} patients with {len(df.columns)} variables")

# Prewritten visualization functions for exploration
def plot_value_counts(series, title, figsize=(10, 6)):
    """
    Create a bar chart of value counts.
    
    Args:
        series: pandas Series with value counts
        title: Chart title
        figsize: Figure size tuple
    """
    plt.figure(figsize=figsize)
    series.plot(kind='bar')
    plt.title(title)
    plt.xticks(rotation=45)
    plt.tight_layout()
    plt.show()

def plot_crosstab(crosstab_data, title, figsize=(10, 6)):
    """
    Create a heatmap of crosstab data.
    
    Args:
        crosstab_data: pandas DataFrame from pd.crosstab()
        title: Chart title
        figsize: Figure size tuple
    """
    plt.figure(figsize=figsize)
    plt.imshow(crosstab_data.values, cmap='Blues', aspect='auto')
    plt.colorbar()
    plt.title(title)
    plt.xticks(range(len(crosstab_data.columns)), crosstab_data.columns, rotation=45)
    plt.yticks(range(len(crosstab_data.index)), crosstab_data.index)
    plt.tight_layout()
    plt.show()

Loaded 10000 patients with 18 variables


## Part 1: Basic Exploration (3 points)

Display:
1. Dataset shape
2. Column names and types
3. First 10 rows
4. Summary statistics (.describe())

In [2]:
# TODO: Display dataset info
display("Dataset shape:", df.shape)
display(f"Columns: {df.columns.tolist()}")
display("Data types:", df.dtypes)
display("First 10 rows:", df.head(10))
display("Summary statistics:", df.describe(include='all'))

'Dataset shape:'

(10000, 18)

"Columns: ['patient_id', 'age', 'sex', 'bmi', 'enrollment_date', 'systolic_bp', 'diastolic_bp', 'cholesterol_total', 'cholesterol_hdl', 'cholesterol_ldl', 'glucose_fasting', 'site', 'intervention_group', 'follow_up_months', 'adverse_events', 'outcome_cvd', 'adherence_pct', 'dropout']"

'Data types:'

patient_id             object
age                     int64
sex                    object
bmi                   float64
enrollment_date        object
systolic_bp           float64
diastolic_bp          float64
cholesterol_total     float64
cholesterol_hdl       float64
cholesterol_ldl       float64
glucose_fasting       float64
site                   object
intervention_group     object
follow_up_months        int64
adverse_events          int64
outcome_cvd            object
adherence_pct         float64
dropout                object
dtype: object

'First 10 rows:'

patient_id  age         sex   bmi enrollment_date  systolic_bp  \
0     P00001   80           F  29.3      2022-05-01        123.0   
1     P00002   80    Female     NaN      2022-01-06        139.0   
2     P00003   82      Female  -1.0      2023-11-04        123.0   
3     P00004   95      Female  25.4      2022-08-15        116.0   
4     P00005   95           M   NaN      2023-04-17         97.0   
5     P00006   78           F  26.8      2023-08-29        116.0   
6     P00007   84           F  25.4      2022-05-12        133.0   
7     P00008   70        Male  24.7      2022-06-04        111.0   
8     P00009   92      Female  26.9      2022-04-06          NaN   
9     P00010   75        Male  21.1      2023-12-14        128.0   

   diastolic_bp  cholesterol_total  cholesterol_hdl  cholesterol_ldl  \
0          80.0              120.0             55.0             41.0   
1          81.0              206.0             58.0            107.0   
2          86.0              172.0             56.0             82.0   
3          77.0              200.0             56.0            104.0   
4          71.0              185.0             78.0             75.0   
5          66.0              164.0             54.0             99.0   
6         100.0              215.0             62.0            113.0   
7          72.0              174.0             60.0             94.0   
8           NaN              189.0             62.0             89.0   
9          76.0              218.0             77.0             97.0   

   glucose_fasting    site intervention_group  follow_up_months  \
0            118.0  site b            Control                20   
1             79.0  Site A            CONTROL                24   
2             77.0  SITE C      treatment b                   2   
3            115.0  Site D        treatment b                17   
4            113.0  site e         Treatmen A                 9   
5             99.0  Site A         TreatmentA                 4   
6             70.0  site a        treatment a                20   
7            109.0  SITE B        TREATMENT A                19   
8            103.0  site a            Control                21   
9             96.0  SITE A        Treatment B                 1   

   adverse_events outcome_cvd  adherence_pct dropout  
0               0          No           24.0      No  
1               0          No           77.0      No  
2               0         Yes           70.0      No  
3               0          No           62.0      No  
4               0         yes            NaN     Yes  
5               0         yes            NaN     Yes  
6               1          No           76.0      No  
7               0          No           53.0      No  
8               0         yes           53.0      No  
9               0          No           50.0      No

'Summary statistics:'

patient_id           age     sex          bmi enrollment_date  \
count       10000  10000.000000   10000  9562.000000           10000   
unique      10000           NaN       8          NaN            1666   
top        P00001           NaN  Female          NaN      2023-06-02   
freq            1           NaN    2684          NaN              23   
mean          NaN     59.182700     NaN    25.730558             NaN   
std           NaN    151.769963     NaN     5.339547             NaN   
min           NaN   -999.000000     NaN    -1.000000             NaN   
25%           NaN     70.000000     NaN    23.500000             NaN   
50%           NaN     80.000000     NaN    26.000000             NaN   
75%           NaN     92.000000     NaN    28.775000             NaN   
max           NaN    100.000000     NaN    42.800000             NaN   

        systolic_bp  diastolic_bp  cholesterol_total  cholesterol_hdl  \
count   9586.000000   9586.000000        9446.000000      9446.000000   
unique          NaN           NaN                NaN              NaN   
top             NaN           NaN                NaN              NaN   
freq            NaN           NaN                NaN              NaN   
mean     117.531087     73.550908         178.039488        61.369786   
std       13.973973     10.167464          33.129034        11.062101   
min       75.000000     60.000000          91.000000        25.000000   
25%      108.000000     65.000000         155.000000        54.000000   
50%      117.000000     73.000000         177.000000        61.000000   
75%      127.000000     81.000000         200.000000        69.000000   
max      173.000000    118.000000         315.000000        98.000000   

        cholesterol_ldl  glucose_fasting    site intervention_group  \
count       9446.000000      9631.000000   10000              10000   
unique              NaN              NaN      34                 26   
top                 NaN              NaN  site b             Contrl   
freq                NaN              NaN     742                802   
mean          85.698603        96.424255     NaN                NaN   
std           28.686463        17.112961     NaN                NaN   
min           40.000000        51.000000     NaN                NaN   
25%           65.000000        84.000000     NaN                NaN   
50%           84.000000        96.000000     NaN                NaN   
75%          105.000000       108.000000     NaN                NaN   
max          226.000000       163.000000     NaN                NaN   

        follow_up_months  adverse_events outcome_cvd  adherence_pct dropout  
count        10000.00000    10000.000000       10000    8533.000000   10000  
unique               NaN             NaN           4            NaN       2  
top                  NaN             NaN          No            NaN      No  
freq                 NaN             NaN        4701            NaN    8533  
mean            12.25460        0.145500         NaN      60.615610     NaN  
std              7.07675        0.393631         NaN      18.974399     NaN  
min              0.00000        0.000000         NaN      20.000000     NaN  
25%              6.00000        0.000000         NaN      47.000000     NaN  
50%             12.00000        0.000000         NaN      62.000000     NaN  
75%             19.00000        0.000000         NaN      75.000000     NaN  
max             24.00000        4.000000         NaN     100.000000     NaN

## Part 2: Column Selection (3 points)

Demonstrate different selection methods:

1. Select only numeric columns using `.select_dtypes()`
2. Select specific columns by name
3. Select a subset of rows and columns using `.loc[]`

In [3]:
# TODO: Select numeric columns
select_numeric = df.select_dtypes(include=[np.number])
display(select_numeric)

age   bmi  systolic_bp  diastolic_bp  cholesterol_total  \
0      80  29.3        123.0          80.0              120.0   
1      80   NaN        139.0          81.0              206.0   
2      82  -1.0        123.0          86.0              172.0   
3      95  25.4        116.0          77.0              200.0   
4      95   NaN         97.0          71.0              185.0   
...   ...   ...          ...           ...                ...   
9995   72  23.2        122.0          73.0              182.0   
9996  100  28.9        124.0          78.0              157.0   
9997   78  23.8        110.0          63.0              154.0   
9998   86  27.0        139.0          98.0              196.0   
9999   67  29.4        134.0          83.0              197.0   

      cholesterol_hdl  cholesterol_ldl  glucose_fasting  follow_up_months  \
0                55.0             41.0            118.0                20   
1                58.0            107.0             79.0                24   
2                56.0             82.0             77.0                 2   
3                56.0            104.0            115.0                17   
4                78.0             75.0            113.0                 9   
...               ...              ...              ...               ...   
9995             54.0             92.0             97.0                21   
9996             56.0             70.0            102.0                11   
9997             69.0             71.0            114.0                 2   
9998             38.0            119.0            126.0                16   
9999             51.0            117.0            128.0                22   

      adverse_events  adherence_pct  
0                  0           24.0  
1                  0           77.0  
2                  0           70.0  
3                  0           62.0  
4                  0            NaN  
...              ...            ...  
9995               0           20.0  
9996               0           57.0  
9997               1           77.0  
9998               0           63.0  
9999               0           48.0  

[10000 rows x 11 columns]

In [4]:
# TODO: Select specific columns
select_specific = df[["age", "bmi", "systolic_bp"]]
display(select_specific)

age   bmi  systolic_bp
0      80  29.3        123.0
1      80   NaN        139.0
2      82  -1.0        123.0
3      95  25.4        116.0
4      95   NaN         97.0
...   ...   ...          ...
9995   72  23.2        122.0
9996  100  28.9        124.0
9997   78  23.8        110.0
9998   86  27.0        139.0
9999   67  29.4        134.0

[10000 rows x 3 columns]

In [5]:
# TODO: Use .loc[] to select subset
subset = df.loc[0:10, 'age':'systolic_bp']
display(subset)

age         sex   bmi enrollment_date  systolic_bp
0    80           F  29.3      2022-05-01        123.0
1    80    Female     NaN      2022-01-06        139.0
2    82      Female  -1.0      2023-11-04        123.0
3    95      Female  25.4      2022-08-15        116.0
4    95           M   NaN      2023-04-17         97.0
5    78           F  26.8      2023-08-29        116.0
6    84           F  25.4      2022-05-12        133.0
7    70        Male  24.7      2022-06-04        111.0
8    92      Female  26.9      2022-04-06          NaN
9    75        Male  21.1      2023-12-14        128.0
10   79      Female  23.5      2023-06-12        110.0

## Part 3: Filtering (4 points)

Filter the data to answer these questions:

1. How many patients are over 65 years old?
2. How many patients have systolic BP > 140?
3. Find patients who are both over 65 AND have systolic BP > 140
4. Find patients from Site A or Site B using `.isin()`

In [6]:
# TODO: Filter and count patients over 65
# 1. Use the filter_data utility from Q3
# 2. Create a filter for age > 65
# 3. Apply the filter and count the results
# print(f"Patients over 65: {len(patients_over_65)}")
import q3_data_utils
filters = [{'column': 'age', 'condition': 'greater_than', 'value': 65}]
patients_over_65 = filter_data(df, filters)
print(f"Patients over 65: {len(patients_over_65)}")

Patients over 65: 8326


In [7]:
# TODO: Filter for high BP
# 1. Use the filter_data utility from Q3
# 2. Create a filter for systolic_bp > 140
# 3. Apply the filter and count the results
# print(f"Patients with high BP: {len(high_bp)}")
import q3_data_utils
filters = [{'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}]
high_bp = filter_data(df, filters)
print(f"Patients with high bp: {len(high_bp)}")

Patients with high bp: 538


In [8]:
# TODO: Multiple conditions with &
# 1. Use filter_data for multiple conditions:
# 2. Create filters for both conditions:
#     {'column': 'age', 'condition': 'greater_than', 'value': 65},
#     {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}
# ]
# 3. Apply the filter and count the results
# print(f"Patients over 65 AND high BP: {len(both_conditions)}")
#
# 5. Alternative: Use in_range for age range:
# 5. Create filter for age range 65-100
# 6. Apply the filter and count the results
import q3_data_utils
filters = [{'column': 'age', 'condition': 'greater_than', 'value': 65},
           {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}]
both_conditions = filter_data(df, filters)
print(f"Patients over 65 AND high BP: {len(both_conditions)}")

Patients over 65 AND high BP: 464


In [9]:
# TODO: Filter by site using .isin()
# 1. Use the filter_data utility from Q3
# 2. Create a filter for Site A or Site B
# 3. Apply the filter and count the results
# print(f"Patients from Site A or B: {len(site_ab)}")
import q3_data_utils
print("Before cleaning:")
print(df['site'].unique())

# Clean the site column
df['site'] = (df['site']
              .str.strip() # removes leading/trailing whitespace
              .str.replace('_', ' ') # replaces all _ with space
              .str.replace(r'\s+', ' ', regex=True) # replaces multiple spaces with single space, r'\s+' is a regex for whitespace
              .str.lower()) # standardizes lowercase

print("After cleaning:")
print(sorted(df['site'].unique()))  # Sorted for easier viewing

# Now filter
filters = [
    {'column': 'site', 'condition': 'in_list', 'value': ['Site A', 'Site B']}
]
site_ab = filter_data(df, filters)
print(f"\nPatients from Site A or B: {len(site_ab)}")  # 4

Before cleaning:
['site b' 'Site A' 'SITE C' 'Site D' 'site e' 'site a' 'SITE B' 'SITE A'
 'SITE E' 'site d' 'Site E' 'site c' 'Site C' '  SITE D  ' '  Site B  '
 'Site_D' 'SITE D' '  site a  ' 'Site B' '  site e  ' '  SITE B  '
 'Site  A' '  Site  A  ' '  site d  ' '  site b  ' '  site c  '
 '  Site E  ' '  SITE E  ' '  SITE C  ' '  SITE A  ' '  Site A  '
 '  Site_D  ' '  Site C  ' '  Site D  ']
After cleaning:
['site a', 'site b', 'site c', 'site d', 'site e']

Patients from Site A or B: 0


## Part 4: Value Counts and Grouping (5 points)

1. Get value counts for the 'site' column
2. Get value counts for the 'intervention_group' column  
3. Create a crosstab of site vs intervention_group
4. Calculate mean age by site
5. Save the site value counts to `output/q4_site_counts.csv`

In [10]:
# TODO: Value counts and analysis
import q3_data_utils
print("Before cleaning:")
print(df['intervention_group'].unique())

# Clean the intervention column
df['intervention_group'] = (df['intervention_group']
              .str.strip() # removes leading/trailing whitespace
              .str.replace('_', ' ') # replaces all _ with space
              .str.replace(r'\s+', ' ', regex=True) # replaces multiple spaces with single space, r'\s+' is a regex for whitespace
              .str.lower()
              .str.replace('contrl', 'control')
              .str.replace('treatmen a', 'treatment a')
              .str.replace('treatmenta', 'treatment a'))

print("After cleaning:")
print(sorted(df['intervention_group'].unique()))  # Sorted for easier viewing

# Get value counts for the 'site' column
site_counts = df['site'].value_counts()
print("\nSite value counts:")
print(f"Site counts: {site_counts}")

# Get value counts for the 'intervention_group' column 
intervention_counts = df['intervention_group'].value_counts()
print("\nIntervention value counts:")
print(f"Intervention counts: {intervention_counts}")

# Create a crosstab of site vs intervention_group
crosstab = pd.crosstab(df['site'], df['intervention_group'])
print("\nCrosstab of site vs intervention group:")
print(crosstab)

# Calculate mean age by site
mean_age_by_site = df.groupby('site')['age'].mean()
print("\nMean age by site:")
print(mean_age_by_site)

df.to_csv('data/clinical_trial_cleaned_column_names.csv')

Before cleaning:
['Control' 'CONTROL' '  treatment b  ' 'treatment b' 'Treatmen A'
 'TreatmentA' 'treatment a' 'TREATMENT A' 'Treatment B' 'control' 'Contrl'
 'Treatment A' 'TREATMENT B' '  control  ' '  TREATMENT A  '
 'Treatment  B' '  TreatmentA  ' '  Contrl  ' '  TREATMENT B  '
 '  treatment a  ' '  Control  ' '  Treatmen A  ' '  Treatment  B  '
 '  Treatment A  ' '  Treatment B  ' '  CONTROL  ']
After cleaning:
['control', 'treatment a', 'treatment b']

Site value counts:
Site counts: site
site a    2956
site b    2453
site c    2073
site d    1501
site e    1017
Name: count, dtype: int64

Intervention value counts:
Intervention counts: intervention_group
treatment b    3342
control        3337
treatment a    3321
Name: count, dtype: int64

Crosstab of site vs intervention group:
intervention_group  control  treatment a  treatment b
site                                                 
site a                 1023          964          969
site b                  780          834  

In [11]:
# TODO: Save output
# site_counts.to_csv('output/q4_site_counts.csv')
site_counts.to_csv('output/q4_site_counts.csv')

## Summary

Write 2-3 sentences about what you learned from exploring this dataset.

**Your summary here:**

TODO: Write your observations

Exploring this dataset, I found that both 'site' and 'intervention_group' columns had data entered that 
varied widely and needed some cleaning of the string format so that I could get accurate counts. 

Additionally, it seems that the value counts across the three intervention groups was pretty evenly split,
with the following results: Treatment B had 3342 value counts, Treatment A had 3321 value counts, and the 
Control had 3337 value counts.
